lasagne/tests/layers/test_autoenc.py

```diff
@@ -0,0 +1,159 @@
+                    b=None)
+    l2 = DenseLayer(l1, num_units=10,
+                    nonlinearity=identity,
+                    b=None)
+    l_ae = build_autoencoder(l2,
+                             nonlinearity='same',
+                             b=None)
+
+    # Check that all layers are the right type
+    autoencoder_layers = get_all_layers(l_ae)
+    assert isinstance(autoencoder_layers[0], InputLayer)
+    assert isinstance(autoencoder_layers[1], DenseLayer)
+    assert isinstance(autoencoder_layers[2], BiasLayer)
+    assert isinstance(autoencoder_layers[3], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[4], DenseLayer)
+    assert isinstance(autoencoder_layers[5], BiasLayer)
+    assert isinstance(autoencoder_layers[6], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[7], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[8], BiasLayer)
+    assert isinstance(autoencoder_layers[9], InverseLayer)
+    assert isinstance(autoencoder_layers[10], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[11], BiasLayer)
+    assert isinstance(autoencoder_layers[12], InverseLayer)
+
+    # Check that the output of the autoencoder is correct
+    ae_output = get_output(l_ae)
+    W1 = l1.W.get_value()
+    W2 = l2.W.get_value()
+    ae_output_fun = theano.function([l_in.input_var], ae_output)
+    X_input = np.random.rand(*input_shape).astype(theano.config.floatX)
+
+    assert np.allclose(
+        ae_output_fun(X_input),
+        np.dot(np.dot(np.dot(np.dot(X_input, W1), W2), W2.T), W1.T),
+        atol=1e-3)
+
+
+def test_unfold_bias_and_nonlinearity_layers():
+    import theano.tensor as T
+    from lasagne.layers import InputLayer, DenseLayer
+    from lasagne.layers import BiasLayer, NonlinearityLayer
+    from lasagne.nonlinearities import tanh, sigmoid, identity
+    from lasagne.layers import get_all_layers, get_output
+    from lasagne.layers import get_all_param_values
+    from lasagne.layers import unfold_bias_and_nonlinearity_layers
+
+    input_shape = tuple(np.random.randint(20, high=100, size=2))
+    l_in = InputLayer(input_shape)
+    l1 = DenseLayer(l_in, num_units=50, nonlinearity=tanh)
+    l_out = DenseLayer(l1, num_units=10, nonlinearity=sigmoid)
+    params_before = get_all_param_values(l_out)
+    output_before = get_output(l_out)
+
+    # add BiasLayers and NonlinearityLayers
+    l_out = unfold_bias_and_nonlinearity_layers(l_out)
+    params_after = get_all_param_values(l_out)
+    output_after = get_output(l_out)
+
+    # Check that all params are the same
+    for p_b, p_a in zip(params_before, params_after):
+        assert np.allclose(p_b, p_a)
+
+    # Check that all layers are the right type,
+    # and that all nonlinearities are correct
+    all_layers = get_all_layers(l_out)
+    assert isinstance(all_layers[0], InputLayer)
+    assert isinstance(all_layers[1], DenseLayer)
+    assert all_layers[1].nonlinearity == identity
+    assert all_layers[1].b is None
```

Comment:
Please also add `assert len(all_layers[1].params) == 1` to ensure that the bias is not part of that layer's learnable parameters any more. Same for the other `DenseLayer` below.